hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardFsck.java

```diff
@@ -0,0 +1,421 @@
+    this.rawFS = fs;
+
+    if (ms == null) {
+      throw new InvalidParameterException("S3AFileSystem should be guarded by"
+          + " a " + DynamoDBMetadataStore.class.getCanonicalName());
+    }
+    this.metadataStore = (DynamoDBMetadataStore) ms;
+
+    if (rawFS.hasMetadataStore()) {
+      throw new InvalidParameterException("Raw fs should not have a "
+          + "metadatastore.");
+    }
+  }
+
+  /**
+   * Compares S3 to MS.
+   * Iterative breadth first walk on the S3 structure from a given root.
+   * Creates a list of pairs (metadata in S3 and in the MetadataStore) where
+   * the consistency or any rule is violated.
+   * Uses {@link S3GuardFsckViolationHandler} to handle violations.
+   * The violations are listed in Enums: {@link Violation}
+   *
+   * @param p the root path to start the traversal
+   * @throws IOException
+   * @return a list of {@link ComparePair}
+   */
+  public List<ComparePair> compareS3ToMs(Path p) throws IOException {
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    int scannedItems = 0;
+
+    final Path rootPath = rawFS.qualify(p);
+    S3AFileStatus root = null;
+    try {
+      root = (S3AFileStatus) rawFS.getFileStatus(rootPath);
+    } catch (AWSBadRequestException e) {
+      throw new IOException(e.getMessage());
+    }
+    final List<ComparePair> comparePairs = new ArrayList<>();
+    final Queue<S3AFileStatus> queue = new ArrayDeque<>();
+    queue.add(root);
+
+    while (!queue.isEmpty()) {
+      final S3AFileStatus currentDir = queue.poll();
+      scannedItems++;
+
+      final Path currentDirPath = currentDir.getPath();
+      List<FileStatus> s3DirListing = Arrays.asList(rawFS.listStatus(currentDirPath));
+
+      // DIRECTORIES
+      // Check directory authoritativeness consistency
+      compareAuthoritativeDirectoryFlag(comparePairs, currentDirPath, s3DirListing);
+      // Add all descendant directory to the queue
+      s3DirListing.stream().filter(pm -> pm.isDirectory())
+              .map(S3AFileStatus.class::cast)
+              .forEach(pm -> queue.add(pm));
+
+      // FILES
+      // check files for consistency
+      final List<S3AFileStatus> children = s3DirListing.stream()
+              .filter(status -> !status.isDirectory())
+              .map(S3AFileStatus.class::cast).collect(toList());
+      final List<ComparePair> compareResult =
+          compareS3DirToMs(currentDir, children).stream()
+              .filter(comparePair -> comparePair.containsViolation())
+              .collect(toList());
+      comparePairs.addAll(compareResult);
+      scannedItems += children.size();
+    }
+    stopwatch.stop();
+
+    // Create a handler and handle each violated pairs
+    S3GuardFsckViolationHandler handler =
+        new S3GuardFsckViolationHandler(rawFS, metadataStore);
+    comparePairs.forEach(handler::handle);
+
+    LOG.info("Total scan time: {}s", stopwatch.elapsed(TimeUnit.SECONDS));
+    LOG.info("Scanned entries: {}", scannedItems);
```

Comment:
I have a little code snippet in cloudstore which dumps all FS stats after some work; this fsck code is somewhere I think it could be useful. Maybe we should have it in S3GuardTool and executed when some -verbose command comes in. https://github.com/steveloughran/cloudstore/blob/master/src/main/java/org/apache/hadoop/fs/store/StoreEntryPoint.java#L317

Reply:
We can add a new jira for a verbose command. I'd like to close this one today, so no more features.

Reply:
added: https://issues.apache.org/jira/browse/HADOOP-16507?focusedCommentId=16926405&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16926405